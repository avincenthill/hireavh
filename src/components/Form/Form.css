.form-input {
  height: var(--s-l);
  font-size: var(--fs-xxs);
  border: var(--c-1) solid var(--b-1);
  border-radius: var(--s-xs);
  color: var(--c-white);
  background-color: var(--c-black);
  margin-bottom: var(--s-s);
  box-sizing: border-box;
  padding: var(--s-xxs);
}

.form-container {
  border: var(--c-1) solid var(--b-1);
  border-radius: var(--s-xs);
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: var(--s-s);
}

.form-error {
  border: var(--c-5) solid var(--b-1);
  color: var(--c-5);
}

.form-item {
  width: min(var(--s-l),65vh);
  display: flex;
  flex-direction: column;
  text-align: left;
  margin-top: var(--s-xs);
}

.form-output {
  text-overflow: wrap;
  width: min(var(--s-xl),80vh);
}

.form-button {
  background-color: var(--c-black);
  border-radius: var(--s-xs);
  border: var(--c-1) solid var(--b-1);
  color: var(--c-1);
  margin: var(--s-xs);
  padding: var(--s-xxs);
}

.form-button:hover {
  background-color: var(--c-4);
  border: var(--c-black) solid var(--b-1);
  color: var(--c-black);
}

.form-item label {
  margin-bottom: var(--s-xxxs);
}

.form-title {
  margin-bottom: var(--s-s);
}