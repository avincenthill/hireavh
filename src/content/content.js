import projects from "./projects/projects";

const currentDate = new Date();

const content = {
  about: {
    subtitle: "👋 Hi, I'm AVH!",
    p1: "Hire me to help build your business with this tech and more!",
  },
  resume: {
    cta: "download 📥",
    downloadName: `Resume_Alex Vincent-Hill_${currentDate.toDateString()}.pdf`,
    path: "assets/pdf/avh_resume.pdf",
  },
  contact: {
    name: `👨‍💻 Alex "AVH" Vincent-Hill`,
    email: "alexanderjvincenthill@gmail.com",
    cell: "(541) 602-9122",
    linkedin: "www.linkedin.com/in/avincenthill",
    github: "www.github.com/avincenthill",
    resume: "resume",
  },
  projects,
  nav: {
    links: [
      { title: "🏠 home", path: "about" },
      { title: "📄 resume", path: "resume" },
      { title: "👨‍💻 contact", path: "contact" },
    ],
  },
  header: {
    title: "hireavh.com 🏠",
  },
  footer: {
    copywrite: `copyright Alex Vincent-Hill ${currentDate.getFullYear()} - 🏠`,
  },
  notFound: {
    title: "🤦 That page is missing, sorry!",
    subtitle: "🏠 return",
  },
};

export default content;
