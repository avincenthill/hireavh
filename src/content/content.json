{
  "title": "hireavh.com",
  "subtitle": "subtitle",
  "paragraph": "paragraph"
}