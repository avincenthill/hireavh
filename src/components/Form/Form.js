import React from "react";
import content from "content/content";
import Page from "components/Page/Page";
import { NavList } from "components/Nav/Nav";
import "./Form.css";

const formContent = content.form;

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      config: formContent.formConfig,
      output: null,
      data: {},
    };
    this.JSONError = formContent.invalidJSONError;
  }

  componentDidMount() {
    // create empty output from placeholder config, check required fields
    this.setOutputData();
    this.setRequiredFields(this.state.config);
  }

  // set the output object from current form config
  setOutputData = () => {
    if (this.state.config && Array.isArray(this.state.config)) {
      const data = {};
      this.state.config.forEach((item) => {
        let element = document.getElementById(item.name);
        if (element.type === "checkbox") {
          data[item.name] = element.checked;
        } else {
          data[item.name] = element.value;
        }
      });
      this.setState({ data });
    }
  };

  // update output object based on current config
  handleInputChange = (event) => {
    const { id, value, checked } = event.target;
    const data = {
      ...this.state.data,
    };
    if (event.target.type === "checkbox") {
      data[id] = checked;
    } else {
      data[id] = value;
    }
    this.setState({ data });
  };

  // push required names into state
  setRequiredFields = (config) => {
    const required = [];
    config.forEach((field) => {
      if (field.is_required) {
        required.push(field.name);
      }
    });
    this.setState({ required });
  };

  // check for valid json and update state on config input change
  handleConfigChange = (event) => {
    const { value } = event.target;
    let config;
    const userInputConfig = value;
    try {
      config = JSON.parse(userInputConfig);
      this.setRequiredFields(config);
    } catch (e) {
      config = this.JSONError;
    }
    this.setState({ config }, this.setOutputData);
  };

  // set output or alerts if required fields values are falsy
  handleSubmit = (event) => {
    event.preventDefault();
    if (this.areRequiredFieldsCompleted()) {
      this.setState({ output: this.state.data });
    } else {
      alert(formContent.requiredWarning + this.state.required.join(", "));
    }
  };

  areRequiredFieldsCompleted = () => {
    let areRequiredFieldsCompleted = true;
    this.state.required.forEach((fieldName) => {
      if (!this.state.data[fieldName]) {
        areRequiredFieldsCompleted = false;
      }
    });
    return areRequiredFieldsCompleted;
  };

  renderForm = (config) => {
    if (config === this.JSONError) {
      return (
        <form className="form-container form-error">
          {formContent.JSONWarning}
        </form>
      );
    } else {
      return (
        <form className="form-container" onSubmit={this.handleSubmit}>
          <h3>{formContent.formTitle}</h3>
          {this.renderFormItems(config)}
          <button className="form-button" type="submit">
            {formContent.submitTitle}
          </button>
        </form>
      );
    }
  };

  renderFormItems = (config) => {
    if (config.length === 0 || !Array.isArray(config)) {
      return <div>{formContent.noItemsWarning}</div>;
    } else {
      return config.map((item, index) => {
        return (
          <div className="form-item" key={index}>
            <label>
              {item.human_label}
              {item.is_required ? "*" : ""}
            </label>
            <input
              id={item.name}
              type={item.type}
              onChange={this.handleInputChange}
            ></input>
          </div>
        );
      });
    }
  };

  render() {
    return (
      <div>
        <Page>
          <h2 className="form-title">{formContent.title}</h2>
          <p className="form-description">{formContent.description}</p>
          <div className="form-item">
            <label>{formContent.inputTitle}</label>
            <textarea
              id="config"
              className="form-input"
              defaultValue={JSON.stringify(formContent.formConfig, null, 2)}
              onChange={this.handleConfigChange}
            ></textarea>
          </div>
          {this.renderForm(this.state.config)}
          <div className="form-item">
            <label>{formContent.outputTitle}</label>
            <textarea
              readOnly
              value={
                this.state.output
                  ? JSON.stringify(this.state.output, null, 2)
                  : formContent.noDataWarning
              }
              className="form-input"
            ></textarea>
          </div>
          <hr></hr>
          <NavList></NavList>
          <hr></hr>
        </Page>
      </div>
    );
  }
}

export default Form;
