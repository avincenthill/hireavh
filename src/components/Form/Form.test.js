import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Form from "./Form";
import content from "content/content";

describe("Form", () => {
  describe("snapshot", () => {
    it("matches the last saved snapshot", () => {
      const tree = renderer.create(<Form />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  }),
    describe("lifecycle", () => {
      let wrapper;
      beforeEach(() => {
        wrapper = shallow(<Form />);
      });
      describe("constructor", () => {
        it("should instantiate with the default config", () => {
          const instance = wrapper.instance();
          expect(instance.state.config).toEqual(content.form.formConfig);
        });
      });
      describe("componentDidMount", () => {
        it("should call setOutputData", () => {
          const instance = wrapper.instance();
          const setOutputDataSpy = jest.spyOn(instance, "setOutputData");
          instance.componentDidMount();
          expect(setOutputDataSpy).toHaveBeenCalledTimes(1);
        });
        it("should call setRequiredFields", () => {
          const instance = wrapper.instance();
          const setOutputDataSpy = jest.spyOn(instance, "setRequiredFields");
          instance.componentDidMount();
          expect(setOutputDataSpy).toHaveBeenCalledTimes(1);
        });
      });
    });
});
