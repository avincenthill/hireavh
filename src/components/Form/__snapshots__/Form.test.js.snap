// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshot should match the last saved snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        className="nav nav-top"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="header-container"
        href="/about"
      >
        <div
          className="header-switch-theme"
        >
          <div
            className="toggle-switch"
          >
            <input
              className="toggle-switch-checkbox"
              defaultChecked={true}
              id="toggle-switch-theme"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="toggle-switch-label"
              htmlFor="toggle-switch-theme"
            >
              <span
                className="toggle-switch-inner"
                data-no="üåô"
                data-yes="‚òÄÔ∏è"
              />
            </label>
          </div>
        </div>
        üè† about
      </a>
    </div>
    <div
      className="page-container"
    >
      <h2
        className="form-title"
      >
        üìÉ Dynamic Form
      </h2>
      <br />
      <p
        className="form-description"
      >
        PERSONAL PROJECT: This form renders inputs based on a configuration array of JSON, for example:
      </p>
      <br />
      <div
        className="form-description"
      >
        [
    {
      "tag": "input",
      "name": "first_name",
      "type": "text",
      "human_label": "First Name",
      "is_required": "true"
    }
  ]
      </div>
      <br />
      <p
        className="form-description"
      >
        It outputs JSON when the form is submitted. This was implemented as a React.js test for an interview and I decided to deploy it due to the frequency I get asked to make something like this while applying for jobs.
      </p>
      <br />
      <div
        className="form-item"
      >
        <label>
          JSON CONFIG
        </label>
        <textarea
          className="form-input"
          defaultValue="[
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"first_name\\",
    \\"type\\": \\"text\\",
    \\"human_label\\": \\"First Name\\",
    \\"is_required\\": \\"true\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"last_name\\",
    \\"type\\": \\"text\\",
    \\"human_label\\": \\"Last Name\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"email\\",
    \\"type\\": \\"email\\",
    \\"human_label\\": \\"Email Address\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"phone_number\\",
    \\"type\\": \\"text\\",
    \\"human_label\\": \\"Phone Number\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"job_title\\",
    \\"type\\": \\"text\\",
    \\"human_label\\": \\"Job Title\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"date_of_birth\\",
    \\"type\\": \\"date\\",
    \\"human_label\\": \\"Date of Birth\\"
  },
  {
    \\"tag\\": \\"input\\",
    \\"name\\": \\"parental_consent\\",
    \\"type\\": \\"checkbox\\",
    \\"human_label\\": \\"Parental Consent\\"
  }
]"
          id="config"
          onChange={[Function]}
        />
      </div>
      <form
        className="form-container"
        onSubmit={[Function]}
      >
        <h3
          className="form-subtitle"
        >
          DYNAMIC FORM
        </h3>
        <div
          className="form-item"
        >
          <label>
            First Name
            *
          </label>
          <input
            id="first_name"
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Last Name
            
          </label>
          <input
            id="last_name"
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Email Address
            
          </label>
          <input
            id="email"
            onChange={[Function]}
            type="email"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Phone Number
            
          </label>
          <input
            id="phone_number"
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Job Title
            
          </label>
          <input
            id="job_title"
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Date of Birth
            
          </label>
          <input
            id="date_of_birth"
            onChange={[Function]}
            type="date"
          />
        </div>
        <div
          className="form-item"
        >
          <label>
            Parental Consent
            
          </label>
          <input
            id="parental_consent"
            onChange={[Function]}
            type="checkbox"
          />
        </div>
        <button
          className="form-button"
          type="submit"
        >
          SUBMIT
        </button>
      </form>
      <div
        className="form-item"
      >
        <label>
          FORM DATA
        </label>
        <textarea
          className="form-input"
          readOnly={true}
          value="ü§î No data returned yet."
        />
      </div>
      <hr />
      <div
        className="nav-list"
      >
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/about"
            >
              <span>
                üè† about
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/blog"
            >
              <span>
                ‚úçÔ∏è blog
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/projects"
            >
              <span>
                üì¶ projects
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/resume"
            >
              <span>
                üìÑ resume
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/contact"
            >
              <span>
                üìá contact
              </span>
            </a>
          </h2>
        </div>
      </div>
      <hr />
    </div>
    <div>
      <div
        className="nav nav-bottom"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="footer-container"
        href="/about"
      >
        <span>
          üè† about
        </span>
      </a>
    </div>
  </div>
</div>
`;
